src/ansys/fluent/core/session.py

```diff
@@ -343,8 +346,87 @@ def download(self, file_name: str, local_file_name: Optional[str] = None):
         FileNotFoundError
             If the remote file does not exist.
         """
-        if self.file_service:
-            if self.file_service.file_exist(file_name):
-                self.file_service.download_file(file_name, local_file_name)
-            else:
-                raise FileNotFoundError("Remote file does not exist.")
+        self._upload_download_helper(
+            is_upload=False, file_name=file_name, local_file_name=local_file_name
+        )
+
+    def _wait_for_file(self, file_name: str):
+        """Wait for file to get ready for upload or download.
+
+        Parameters
+        ----------
+        file_name : str
+            File name
+        Raises
+        ------
+        FileNotFoundError
+            If a file does not exist.
+        """
+        start_time = time.time()
+        max_wait_time = sys.maxsize
+        while (time.time() - start_time) < max_wait_time:
+            if self.file_service.file_exist(os.path.basename(file_name)):
+                break
+            max_wait_time -= 1
+            time.sleep(3)
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+
+    def _pypim_upload(self, file_name: str, on_uploaded: Optional[Any] = None):
+        """Uploads a file if not available on the server.
+
+        Parameters
+        ----------
+        file_name : str
+            File name
+        on_uploaded: Callable[str]
+            Read a file.
+        Raises
+        ------
+        FileNotFoundError
+            If a file does not exist.
+        """
+        if os.path.isfile(file_name):
+            if not self.file_service.file_exist(os.path.basename(file_name)):
+                self.upload(file_name)
+                self._wait_for_file(file_name)
+        elif self.file_service.file_exist(os.path.basename(file_name)):
+            pass
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+        if on_uploaded:
+            on_uploaded(os.path.basename(file_name))
+
+    def _pypim_download(self, file_name: str, api: Optional[Any] = None):
+        """Downloads a file from the server.
+
+        Parameters
+        ----------
+        file_name : str
+            File name
+        api: Callable[str]
+            Write a file.
+        """
+        if api:
+            api(os.path.basename(file_name))
+        self._wait_for_file(file_name)
+        if os.path.isfile(file_name):
+            print(f"\nFile already exists. File path:\n{file_name}\n")
+        else:
+            self.download(os.path.basename(file_name), ".")
+
+    def _no_pypim_helper(
+        self,
+        file_name: str,
+        api: Optional[Any] = None,
+    ):
+        """Used if pypim is not configured.
+
+        Parameters
+        ----------
+        file_name : str
+            File name
+        api: Callable[str]
+            Read or write a file.
+        """
+        api(file_name)
```

Comment:
- We don't seem to need any _no_pypim code as it's simply calling the callback
- As discussed, please update the callback names (not `api` but `on_uploaded` etc)

Reply:
* Removed `_no_pypim` code.
*  Renamed `api` to `on_uploaded` 

Thank you.